### PR TITLE
SKA-26: Unit Tests after Refactoring

### PR DESCRIPTION
**Pull request by rohan-sahgal**
# Description

Completed and refactored test cases covering our existing OpenWeatherService functionality.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Tested by running Pytest locally, all tests pass.

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my code.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by rohan-sahgal** [comment]
Lots of mypy errors during commit - couldn't really figure out how to fix:

`black....................................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

tests/test_openweather_service.py:119: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:122: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:146: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:149: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:194: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:197: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:199: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:221: error: Incompatible types in assignment (expression has type "MoonData", variable has type "MockOpenweatherResponse")  [assignment]
tests/test_openweather_service.py:225: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:228: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:231: error: "MockOpenweatherResponse" has no attribute "moonrise"  [attr-defined]
tests/test_openweather_service.py:232: error: "MockOpenweatherResponse" has no attribute "moonset"  [attr-defined]
tests/test_openweather_service.py:234: error: "MockOpenweatherResponse" has no attribute "moonphase"  [attr-defined]
tests/test_openweather_service.py:234: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
tests/test_openweather_service.py:306: error: No return value expected  [return-value]
tests/test_openweather_service.py:311: error: No return value expected  [return-value]
Found 16 errors in 1 file (checked 2 source files)`

Skipped mypy for initial commit

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -8,4 +8,4 @@ repos:
     hooks:
     -   id: mypy
         args: [--strict, --ignore-missing-imports]
-        additional_dependencies: [types-requests==2.31.0]
\ No newline at end of file
+        additional_dependencies: [types-requests==2.31.0, pydantic==2.5.3]
```
Comment (thereisnoaddress) [inline .pre-commit-config.yaml:11]: no worries about not needing a caret here -- thanks for including this!

File: sky_alert/protocol.py [hunk 2]
```diff
@@ -1,36 +1,36 @@
...
 
-class OpenweatherResponse(BaseModel):  # type: ignore
+class OpenweatherResponse(BaseModel):
     """Class for generic responses"""
 
     status_code: int
     message: str
 
-    @validator("status_code")
+    @field_validator("status_code")
```
Comment (thereisnoaddress) [inline sky_alert/protocol.py:33]: 🚀

File: tests/test_openweather_service.py [hunk 3]
```diff
@@ -254,12 +288,40 @@ def test_given_moonrise_moonset_without_data(self, mock_get: Any) -> None:
...
+
+        # WHEN
+        response = self.ows.get_cloud_data(lat=sample_lat, lon=sample_lon)
+
+        # THEN
+        cloud_data = []
+        for i in range(HOURS_IN_DAY):
+            cloud_data.append(MOCK_OPENWEATHER_RESPONSE_JSON["hourly"][i]["clouds"])
+
+        self.assertEqual(response.cloud_cover, cloud_data)
```
Comment (thereisnoaddress) [inline tests/test_openweather_service.py:318]: nice!

**Review by thereisnoaddress — APPROVED** [review]
Great job understanding and adding these tests! Very impressed; very good job 👏

**Comment by thereisnoaddress** [comment]
> Lots of mypy errors during commit - couldn't really figure out how to fix:
> 
> ```
> black....................................................................Passed
> mypy.....................................................................Failed
> - hook id: mypy
> - exit code: 1
> 
> tests/test_openweather_service.py:119: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:122: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:146: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:149: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:194: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:197: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:199: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:221: error: Incompatible types in assignment (expression has type "MoonData", variable has type "MockOpenweatherResponse")  [assignment]
> tests/test_openweather_service.py:225: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:228: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:231: error: "MockOpenweatherResponse" has no attribute "moonrise"  [attr-defined]
> tests/test_openweather_service.py:232: error: "MockOpenweatherResponse" has no attribute "moonset"  [attr-defined]
> tests/test_openweather_service.py:234: error: "MockOpenweatherResponse" has no attribute "moonphase"  [attr-defined]
> tests/test_openweather_service.py:234: error: Value of type "Collection[Collection[str]]" is not indexable  [index]
> tests/test_openweather_service.py:306: error: No return value expected  [return-value]
> tests/test_openweather_service.py:311: error: No return value expected  [return-value]
> Found 16 errors in 1 file (checked 2 source files)
> ```
> 
> Skipping for now - will deal later

I think a bunch of these can be resolved by adding this to the top of `test_constants.py`: 

hat tip to: https://github.com/python/mypy/issues/8277#issuecomment-1572806494

```
from mypy.types import JsonDict

MOCK_OPENWEATHER_RESPONSE_JSON: JsonDict = {
```

File: .pre-commit-config.yaml (hunk 1 above)
Comment (rohan-sahgal) [reply .pre-commit-config.yaml:11]: no worries 👍

File: tests/test_openweather_service.py (hunk 3 above)
Comment (rohan-sahgal) [reply tests/test_openweather_service.py:318]: no you

**Pull request merged**